DNN-based Tau-Id discrimians

mbluj opened the pull request:
This pull request provides two new DNN-based Tau-Ids, DeepTau and DPFTau to be produced for pat::Taus with MiniAOD.

Details:
* Two new `edm::stream::EDProducer`s called `DeepTauId` and `DPFIsolation` with a common base class `DeepTauBase`,
* DNN algorithms with TensorFlow library,
* Training files with a separate repository https://github.com/cms-tau-pog/RecoTauTag-TrainingFiles forked from https://github.com/cms-data/RecoTauTag-TrainingFiles. PR to the latter will follow this PR. Installation recipe for tests below,
* Python tool providing updated pat::Tau collection and an example configuration.

The new Tau-Ids are not part of any current workflow and primary they are meant to be run within users' workflows. Their integration as a part of NanoAOD or/and MiniAOD will be a subject of separate pull requests when this one is accepted. 
Backports to 94X and 102X release series (used for 2016/17 and 2018 datasets) is foreseen when  this one is accepted.

Recipe for tests (as new modules are not part of standard workflows):
* Checkout needed training files:
```
mkdir -p $CMSSW_BASE/external/$SCRAM_ARCH/data
cd $CMSSW_BASE/external/$SCRAM_ARCH/data
git clone https://github.com/cms-tau-pog/RecoTauTag-TrainingFiles -b master RecoTauTag/TrainingFiles/data
cd -
```
* Edit and run tests configuration:
```
cmsRun RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py
```

kandrosov committed:
First implementation of deep tau id.

ocolegrove committed:
Building dpf isolation module

ocolegrove committed:
Adding in v1

ocolegrove committed:
Adding in runTauIDMVA for other users

ocolegrove committed:
making things fully reproducible

mbluj committed:
Reorganisation of configuration files: cff split to cfi and cff

mbluj committed:
Some code cleaning

mbluj committed:
adapt to cfi/cff reorganization

MRD2F committed:
Review of DPF and DeepTauId code.

- Defined base class for deep tau discriminators.
- Removed weight files from home cms repository. Now using weights from cms-data.
- Defined WP for both discriminators. Now all discriminators return the corresponding WP results.
- Removed cfi files. Using fillDescriptions instead.
- General code review and cleaning.

MRD2F committed:
Added example of a python configuration file to produce pat::Tau collection with the new Tau-Ids

MRD2F committed:
requested changes on runDeepTauIDsOnMiniAOD.py

mbluj committed:
Clean runTauIdMVA.py tool and test config to run tauIDs

MRD2F committed:
Fixed problems to be compatible with branch CMSSW_10_4_X

MRD2F committed:
Made DeepTauId and DPFIsolation thread-safe

MRD2F committed:
Finish implement thread-safe requirements on DPFIsolation

MRD2F committed:
Remove trace of TauIdMVAAuxiliaries in DeepTauId

MRD2F committed:
Disable DPFTau_2016_v1 and issue some warnings

mbluj committed:
Merge pull request #100 from MRD2F/CMSSW_10_4_X_DPFIsoDeepTau

DNN tau IDs for CMSSW_10_4_X

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25016/6991

cmsbuild commented:
A new Pull Request was created by  @mbluj  for master.

It involves the following packages:

RecoTauTag/RecoTau


@perrotta, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@mbluj 
please provide some info about the CPU and memory use.
I see that there are 150MB of training files, which looks a bit scary (IIRC, the size use in memory is typically larger).
see e.g. https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#pp_MC_Benchmark_setup_for_timing
" Checking memory use with IgProfService "

mbluj commented:
> please provide some info about the CPU and memory use.
> I see that there are 150MB of training files, which looks a bit scary
> (IIRC, the size use in memory is typically larger).
> see e.g.
> https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#pp_MC_Benchmark_setup_for_timing
> " Checking memory use with IgProfService "
>
Sure, will be done.

mbluj commented:
@slava77 A couple of naive questions: 
* should I perform the IgProof tests with 1- or multi- (how many?) threaded run?
* ttbar is benchmark sample, right?
* how many events, 100, 1k more?
Thanks.

mbluj commented:
I attach results of igproof profiling with 100 of ttbar events with PU using 103X RelVal. The run was 1 threaded with DPFTau and DeepTau tau-Ids,  tau-Id embedded module + "services" (source, output module, etc.) 
[runDeepTauIDsOnMiniAOD_IgProf_1.txt.gz](https://github.com/cms-sw/cmssw/files/2519850/runDeepTauIDsOnMiniAOD_IgProf_1.txt.gz)
[runDeepTauIDsOnMiniAOD_igCPU.txt.gz](https://github.com/cms-sw/cmssw/files/2519851/runDeepTauIDsOnMiniAOD_igCPU.txt.gz)

slava77 commented:
On 10/26/18 5:41 PM, mbluj wrote:
> I attach results of igproof profiling with 100 of ttbar events with PU
> using 103X RelVal. The run was 1 threaded with DPFTau and DeepTau
> tau-Ids, tau-Id embedded module + "services" (source, output module, etc.)
> runDeepTauIDsOnMiniAOD_IgProf_1.txt.gz

only a quick lookup:

595'443'772        157'182  deep_tau::DeepTauBase::DeepTauBase
This is 600MB of allocations.
I think that at the moment it's likely a no-go.
We will have to have a very good justification to make this into
production as is
(with some good effort put to showing that essentially the same
performance can't be achieved by a simpler model)

> runDeepTauIDsOnMiniAOD_igCPU.txt.gz

>

slava77 commented:
On 10/26/18 6:01 PM, Slava Krutelyov wrote:
> On 10/26/18 5:41 PM, mbluj wrote:
>> I attach results of igproof profiling with 100 of ttbar events with PU
>> using 103X RelVal. The run was 1 threaded with DPFTau and DeepTau
>> tau-Ids, tau-Id embedded module + "services" (source, output module, etc.)
>> runDeepTauIDsOnMiniAOD_IgProf_1.txt.gz
> 
> only a quick lookup:
> 
> 595'443'772 157'182 deep_tau::DeepTauBase::DeepTauBase
> This is 600MB of allocations.
> I think that at the moment it's likely a no-go.

on top of this, there is also
256MB in deep_tau::DeepTauBase::produce
also, a problem

> We will have to have a very good justification to make this into
> production as is
> (with some good effort put to showing that essentially the same
> performance can't be achieved by a simpler model)
> 
>> runDeepTauIDsOnMiniAOD_igCPU.txt.gz
> 
>>
> 
>

mbluj commented:
There are two different modules inheriting from DeepTauBase and I will profile them separately, but after this weekend.
@kandrosov, @ocolegrove, @steggema, @rmanzoni please follow this discussion if not done, yet.

kandrosov commented:
Hi @slava77 
> 595'443'772 157'182 deep_tau::DeepTauBase::DeepTauBase This is 600MB of allocations. I think that at the moment it's likely a no-go.

From a quick check, I obtain that each of the two discriminators contributes almost equaly into 600MB:
343'061'309 - run with DPFTau_2016_v0 only
257'416'319 - run with deepTau2017v1 only

> on top of this, there is also 256MB in deep_tau::DeepTauBase::produce also, a problem

Memory consumption in the produce method can easily be reduced by factor 10. The only drawback is that it'll make code less "gpu-friendly", because each tau will be processed one-by-one, while now they are processed all together, but it is not an issue for the moment, I suppose.

> We will have to have a very good justification to make this into production as is (with some good effort put to showing that essentially the same performance can't be achieved by a simpler model) 

Could you please, give use an idea about what would be a "limit-reasonable" memory consum (order of) that would not require "some good effort" to demonstrate that it is the necessary minimum, below which some significan performance sacrifices should be made?

slava77 commented:
On 10/26/18 8:45 PM, Konstantin Androsov wrote:
> Could you please, give use an idea about what would be a
> "limit-reasonable" memory consum (order of) that would not require "some
> good effort" to demonstrate that it is the necessary minimum, below
> which some significan performance sacrifices should be made?

For this case I can semi-randomly name a 20MB total, below which it
would probably not be worth much  effort beyond just making sure that
the implementation is not obviously inflated.

perrotta reviewed (commented):
A few sparse comment after a first quick glance at the code.
Others will likely follow as soon as you resolve the memory footprint issue.

mbluj commented:
@MRD2F, please check comments on code style given in this thread, thanks.

MRD2F committed:
- Implemented on runTauIdMVA the option to work with new training files quantized
- Added a new parameter 'version' on runTauIdMVA, used on DPFIsolation
- Changes on DeepTauId to reduce memory consumption

MRD2F committed:
- Implementation of global cache to avoid reloading graph for each thread and reduce the memory consuption
- Creation of class DeepTauCache in DeepTauBase, in which now is created graph and session
- Implementation of two new static methods inside the class DeepTauBase: initializeGlobalCache and globalEndJob. The graph and DeepTauCache object are created now inside initializeGlobalCache

kandrosov committed:
Implemented TauWPThreshold class.

TauWPThreshold class parses WP cut string (or value) provided in the
python configuration. It is needed because the use of the standard
StringObjectFunction class to parse complex expression results in an
extensive memory usage (> 100 MB per expression).

MRD2F committed:
Applied changes on DeepTauBase to allow load new training files using memory mapping

MRD2F committed:
Remove the qm.pb input files and leaving just the quantized and the original files

MRD2F committed:
-Overall, changes to improve memory usage, among these are:
	- Implementation of global cache to avoid reloading graph for each thread
	- Creation of two new static methods inside the class DeepTauBase: initializeGlobalCache and globalEndJob. The graph and DeepTauCache object are created now inside initializeGlobalCache. The memory consumption of initializeGlobalCache for the original, quantized and files that are load using memory mapping method are in the memory_usage.pdf file
	- Implemented configuration to use new training files quantized, and set them as default
	- Implementation of configuration for load files using memory mapping. In our case there wasn't any improvement, respect at the memory consumption of this method, respect the quantized files, so this is not used, but set for future training files
- General code review and cleaning.

MRD2F committed:
Applied style comments

MRD2F committed:
Applied style comments

MRD2F committed:
Applied comments

MRD2F committed:
Change to be by default the original training file for deepTau, instead of the quantized

MRD2F committed:
Applied commets of previus PR

MRD2F committed:
cleaning code

MRD2F committed:
Modification in the config to work with new label in files

MRD2F committed:
Applied comment about the expected format of name of training file

MRD2F committed:
Fix in last commit

MRD2F committed:
Applied last comments

MRD2F committed:
Changes regarding forward-porting DNN-related developments from the PRs #105 and #106 from 94X to 104X

mbluj committed:
Merge pull request #107 from MRD2F/CMSSW_10_4_0_pre2_DNN

forward-porting DNN-related from branch cms-tau-pog:CMSSW_9_4_X_tau_pog_DNNTauIDs

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25016/7322

cmsbuild commented:
Pull request #25016 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

mbluj commented:
This PR was updated to address issues discussed above; namely the update contains:
* Changes to process tau candidates one-by-one rather than all together (GPU-friendly => CPU-friendly processing);
* Implementation of a global cache to avoid reloading training graphs for each thread;
* Quantization of training-files to shrink their size => size reduction by a factor of about 3.8 (updated related PR in data repository: https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/1);
* Implementation (configurable, i.e. with possibility to switch on/off) training file load using memory
mapping. It does not give any improvement now concerning the memory consumption, but set for future training files and improvements expected for newer TensorFlow (currently off);
* Custom parser for WP definitions to replace StringObjectFunction which is expensive in memory usage terms;
* Code style cleaning.

The changes gives memory reduction by a factor of about 4, mostly thanks to to reduced size of
training files. Detailed measures can be found in the following thread: https://github.com/cms-tau-pog/cmssw/pull/107.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

mbluj commented:
> please test

Just for notice: the new developments neither modify existing modules nor add anything to existing workflows. Therefore, it is unexpected that standard test will reveal any differences. 
Recipe to test the developments is given in the initial message describing this PR.

cmsbuild commented:
-1

Tested at: 1f7811388d37d03f078e6da39fa54eb88d438478


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/1f7811388d37d03f078e6da39fa54eb88d438478
https://github.com/cms-sw/cmssw/commit/9173af878af87638210fd7e21f8cd65c8190ceb7
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31757/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31757/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31757/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 16 COMPILER='llvm compile'

```
 See details on the summary page.


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/1f7811388d37d03f078e6da39fa54eb88d438478
https://github.com/cms-sw/cmssw/commit/9173af878af87638210fd7e21f8cd65c8190ceb7
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31757/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31757/git-merge-result

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

mbluj committed:
Remove assigning value of variable to itself

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25016/7324

cmsbuild commented:
Pull request #25016 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 3a721da43d52028fda79b3d85d45c039d3d8e5a3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31764/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/1f7811388d37d03f078e6da39fa54eb88d438478
https://github.com/cms-sw/cmssw/commit/9173af878af87638210fd7e21f8cd65c8190ceb7
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31764/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31764/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31764/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3013311
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3013111
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 134 log files, 14 edm output root files, 32 DQM output files

perrotta commented on RecoTauTag/RecoTau/plugins/DPFIsolation.cc:
```diff
@@ -0,0 +1,387 @@
+public:
+    static OutputCollection& GetOutputs()
+    {
+        static size_t tau_index = 0;
```
static const (here and below)

perrotta commented on RecoTauTag/RecoTau/plugins/DPFIsolation.cc:
```diff
@@ -0,0 +1,387 @@
+
+                pfCandTauIndMatch = pfCandTauIndMatch_temp;
+                pfCandPtRelPtRel = pfCandPtRel*pfCandPtRel;
+                if (pfCandPt > 500) pfCandPt = 500.;
```
Is it ok to hard-code all these boundaries as such? (here and below)

perrotta reviewed (commented):
You still need to integrate some of the past comments of mine about the code.
When you do so, please take also into account the ones here below.

mbluj commented:
@MRD2F @kandrosov could you take a look, please?

slava77 commented on RecoTauTag/RecoTau/plugins/DPFIsolation.cc:
```diff
@@ -0,0 +1,387 @@
+                if (pfCandPtRelPtRel > 1) pfCandPtRelPtRel = 1;
+
+                if (pfCandD0 > 5.) pfCandD0 = 5.;
+                if (pfCandD0 < -5.) pfCandD0 = -5.;
```
btw, this fits in one line these days ```pfCandD0 = std::clamp(pfCandD0, -5., 5.);```

MRD2F commented on RecoTauTag/RecoTau/plugins/DPFIsolation.cc:
```diff
@@ -0,0 +1,387 @@
+                    pfCandLostInnerHits = p.lostInnerHits();
+                } else {
+                    float disp = 1;
+                    int psudorand = p.pt()*1000000;
```
@ocolegrove please comment

mbluj replied:
In my understanding it is that DNN prefers that some values are always set for each variable even if the values have not any meaning for a given category, here track parameters for particles w/o tracks. In this particular case, the values are set to values on the boundaries of allowed ranges, i.e. either at min or max value. Some variables (here IPs) are signed thus sign is added at random basis to not prefer negative or positive boundary. After said this, at least naivety IP's values set to zero should also work, however it could require retraining. @ocolegrove, am I right?

MRD2F committed:
Applied @perrotta comments on 104X

MRD2F committed:
Fix error

MRD2F committed:
Applied comments

MRD2F committed:
Applied comments

MRD2F committed:
Merge remote-tracking branch 'tauPOG/CMSSW_10_4_X_tau_pog_DNNTauIDs' into CMSSW_10_4_0_pre2_DNN

MRD2F committed:
Fix merge problem

MRD2F committed:
Applied a few commets

MRD2F committed:
Applied more changes

MRD2F committed:
Applied a few small followups

MRD2F committed:
Fixed error on DPFIsolation

mbluj committed:
Merge pull request #109 from MRD2F/CMSSW_10_4_0_pre2_DNN

Applied comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25016/7348

cmsbuild commented:
Pull request #25016 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

mbluj committed:
Update DPFIsolation.cc

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25016/7349

cmsbuild commented:
Pull request #25016 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 77c09c7f7a29e205d814e9636511ee6ff46bf376
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31813/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/61321279ee46f369212992e86e9d6f9af46e48c9
https://github.com/cms-sw/cmssw/commit/a27ec47d770bb2aaef8ffb873982f5b43da49e77
https://github.com/cms-sw/cmssw/commit/1f7811388d37d03f078e6da39fa54eb88d438478
https://github.com/cms-sw/cmssw/commit/9173af878af87638210fd7e21f8cd65c8190ceb7
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31813/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31813/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/31813/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3131839
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3131632
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

mbluj commented:
Hello, 
are there any further issues with this development or it can be merged (together with related training files) for analysts use (i.e. outside official workflows)? 
What about possible backports to 94X/102X?
Thanks

perrotta commented:
@mbluj : I run the test according to the recipe described in the PR description, and I end up with a ```patTuple_newTauIDs.root``` output file, whose content is
```
Type                  Module               Label     Process   
---------------------------------------------------------------
vector<pat::Tau>      "slimmedTausNewID"   ""        "TauID"   
```

When I try to open it with a root TBrowser and inspect its content, everything hangs forever, and I have to crash the process. Do you have any suggestion for having a look at that file content?

mbluj commented:
This file contains just a new collection of `pat::Taus` containing the new DNN discriminants. It could be that TBrowser has problem with solving Ref's to not-stored collections, however it sounds strange. I'll check it. In meanwhile you can try to run with `minimalOutput = True` which will store original MiniAOD plus the new `pat::Tau` collection.

mbluj commented:
I've just reproduced your issue also with full MiniAOD with `minimalOutput = False`. It looks like an issue with TBrowser. 
When the new collection is accessed directly, e.g. with TTree::Draw it works fine, e.g.
```
Events->Draw("patTaus_slimmedTausNewID__TauID.obj.pt()","");
Events->Draw("patTaus_slimmedTausNewID__TauID.obj.tauID(\"byDpfTau2016v0VSallraw\")","");
std::vector<std::string> idNames = {"byDeepTau2017v1VSeraw", "byDeepTau2017v1VSjetraw", "byDeepTau2017v1VSmuraw", "byDpfTau2016v0VSallraw"};
for(const auto & idName: idNames) {
   Events->Draw((std::string("patTaus_slimmedTausNewID__TauID.obj.tauID(\"")+idName+std::string("\")")).c_str(),"");
}
```

perrotta commented:
+1
- New DNN-based Tau-Id discriminants are made available for analyses: 
   - they don't run in official central workflows
   - additional training files are also needed
- Memory footprint is still rather large (O(100MB/evt)) , but nonetheless deemed acceptable for the private usage this code is intended to: if and when it will be requested to be included in some central sequence meant to production some further assessment and study should be performed, as well as some deeper investigation of the timing performance (which from the first results does not seem to be a concern, anyhow)
- Since central sequences used in production are not touched by this PR, its possible backport has no counterindications
- Jenkins tests pass and show no differences

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
@mbluj @perrotta @slava77 the test provided runs smoothly once the training file is downloaded. The PR in itself looks ready to go, and as it is not supposed to run in production workflows for the time being it could be ok. Nevertheless I see https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/1 and cms-sw/cmsdist#4455 . I understand they are ok to be merged and work with this code, am I correct? 

@smuzaffar FYI

mbluj commented:
> [...]
> Nevertheless I see [cms-data/RecoTauTag-TrainingFiles#1](https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/1) and [cms-sw/cmsdist#4455](https://github.com/cms-sw/cmsdist/issues/4455) . I understand they are ok to be merged and work with this code, am I correct?

Yes, I confirm: subject of [cms-data/RecoTauTag-TrainingFiles#1](https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/1) and the related cmsdist issue are training files needed for the DNN-tau-Ids; those files are identical as used for the tests.

fabiocos commented:
please test with cms-sw/cmsdist#4554

cmsbuild commented:
The tests are being triggered in jenkins.
Using externals from cms-sw/cmsdist#4554

smuzaffar commented:
please test with cms-w/cmsdist#4554

cmsbuild commented:
+1
Tested at: 77c09c7f7a29e205d814e9636511ee6ff46bf376
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/32002/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/61321279ee46f369212992e86e9d6f9af46e48c9
https://github.com/cms-sw/cmssw/commit/a27ec47d770bb2aaef8ffb873982f5b43da49e77
https://github.com/cms-sw/cmssw/commit/1f7811388d37d03f078e6da39fa54eb88d438478
https://github.com/cms-sw/cmssw/commit/9173af878af87638210fd7e21f8cd65c8190ceb7
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/32002/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/32002/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25016/32002/summary.html

The workflows 1001.0, 1000.0, 140.53, 136.85, 136.8311, 136.7611, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3131939
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3131734
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

fabiocos commented:
+1
